cupy/linalg/einsum.py

```diff
@@ -64,22 +77,57 @@ def calc_summed_view(ioperand, input_subscript, output_subscript):
     assert len(set(output_subscript)) == len(output_subscript)
     assert set(output_subscript).issubset(set(input_subscript))
 
-    subscript = input_subscript
-    label_to_summed = set(input_subscript) - set(output_subscript)
+    input_subscript_excluded_at = input_subscript.replace('@', '')
+
+    label_to_summed = set(input_subscript_excluded_at) - set(output_subscript)
     axes_to_summed = []
-    for i, label in enumerate(input_subscript):
+    ellipsis_pos = input_subscript.find('@')
+    for i, label in enumerate(input_subscript_excluded_at):
         if label in label_to_summed:
-            axes_to_summed.append(i)
+            if ellipsis_pos == -1 or i < ellipsis_pos:
+                axes_to_summed.append(i)
+            else:
+                axes_to_summed.append(i - len(input_subscript_excluded_at))
 
     if axes_to_summed:
         result = ioperand.sum(axis=tuple(axes_to_summed)). \
-            astype(ioperand)
+            astype(ioperand.dtype)
     else:
         result = ioperand
     for label in label_to_summed:
-        subscript = subscript.replace(label, '')
+        input_subscript = input_subscript.replace(label, '')
 
-    return result, subscript
+    return result, input_subscript
+
+
+def _normalize_axis_tuple(axis, ndim):
+    """Normalizes an axis argument into a tuple of non-negative integer axes.
+
+    .. seealso:: :func:`numpy.core.numeric.normalize_axis_tuple`
+    """
+    return tuple((ax % ndim) for ax in axis)
```

Comment:
Is it safe when `ax` is larger than `2 * ndim`?

Reply:
I added exception if ax > ndim as `numpy.normalize_axis_tuple`.